File: s3.ml
(**  Get the S3 class of a given SEXP.
  *
  *  r_s3_class takes a SEXP as argument, and returns the S3 class
  *  attribute of the given SEXP.
  *)

external s3_class : sexp -> sexp = "r_s3_class"

external aux_get_attrib : sexp -> sexp -> sexp = "r_get_attrib"
let get_attrib s name = aux_get_attrib s (install name)

external get_attributes : sexp -> sexp = "r_get_attributes"

module S3 = struct

  class type t = object

    val underlying    : sexp
    method underlying : sexp
    method attribute  : string -> sexp
    method attributes : sexp
    method classes    : string list

  end

  class from_R r : t = object

    val underlying = r
    method underlying = underlying
    method attribute s = get_attrib underlying s
    method attributes = get_attributes underlying
    method classes = strings_of_t (get_attrib underlying "class")

  end

  let t_from_R r : t = new from_R r

end

File: s3.c
/**  Get the S3 class of a given SEXP.
  *
  *  r_s3_class takes a SEXP as argument, and returns the S3 class
  *  attribute of the given SEXP.
  *
  *  This calls getAttrib, which is part of the API, to access the
  *  "class" attribute. Usage is (weakly) documented in section
  *  5.9.4 "Attributes" of the "Writing R extensions" documentation,
  *  a.k.a. r-exts.pdf.
  *
  *  Note: Calls to getAttrib do a lot of unnecessary dynamic type
  *  checking. This will have to clarified and cleaned up in future
  *  versions of OCaml-R, possibly bypassing the API.
  */

CAMLprim value r_s3_class (value sexp) {
  CAMLparam1(sexp);
  CAMLlocal1(result);
  result = Val_sexp(getAttrib(Sexp_val(sexp), R_ClassSymbol));
  CAMLreturn(result);
}


/**  Get an attribute of a given SEXP.
  *
  *  r_get_attrib takes a SEXP as first argument, then an R symbol
  *  name, and returns the attribute of the first argument matching
  *  this symbol name. Such symbols can be created with the install
  *  function, from Objective Caml side.
  *
  *  Note: As above, lots a dynamic type checking that will have to
  *  be bypassed.
  */

CAMLprim value r_get_attrib (value sexp, value symbolname) {
  CAMLparam1(sexp);
  CAMLlocal1(result);
  result = Val_sexp(getAttrib(Sexp_val(sexp), Sexp_val(symbolname)));
  CAMLreturn(result);
}


/**  Get the attribute list of a given SEXP.
  *
  *  r_get_attributes takes a SEXP as first argument, and returns
  *  the attributes of the given SEXP.
  */

CAMLprim value r_get_attributes (value sexp) {
  CAMLparam1(sexp);
  CAMLlocal1(result);
  result = Val_sexp(ATTRIB(Sexp_val(sexp)));
  CAMLreturn(result);
}